clusterloader2/pkg/test/simpletestexecutor.go

```diff
@@ -27,21 +35,142 @@ func createSimpleTestExecutor() TestExecutor {
 }
 
 // ExecuteTest executes test based on provided configuration.
-func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) []error {
+	defer cleanupResources(ctx)
+	var errList []error
+	err := ctx.GetFramework().CreateAutomanagedNamespaces(int(conf.AutomanagedNamespaces))
+	if err != nil {
+		errList = append(errList, fmt.Errorf("automanaged namespaces creation failed: %v", err))
+		return errList
+	}
+
+	for i := range conf.Steps {
+		if stepErrList := ste.ExecuteStep(ctx, &conf.Steps[i]); len(stepErrList) > 0 {
+			errList = append(errList, stepErrList...)
+			if isErrsCritical(stepErrList) {
+				return errList
+			}
+		}
+	}
+	return errList
 }
 
 // ExecuteStep executes single test step based on provided step configuration.
-func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) []error {
+	var errList []error
+	if len(step.Measurements) > 0 {
+		// TODO: handle measurments
+	} else {
+		var wg wait.Group
+		var lock sync.Mutex
+		for i := range step.Phases {
+			phase := &step.Phases[i]
+			wg.Start(func() {
+				if phaseErrList := ste.ExecutePhase(ctx, phase); len(phaseErrList) > 0 {
+					lock.Lock()
+					defer lock.Unlock()
+					errList = append(errList, phaseErrList...)
+				}
+			})
+		}
+		wg.Wait()
+	}
+	return errList
 }
 
 // ExecutePhase executes single test phase based on provided phase configuration.
-func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) error {
+func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) []error {
+	var errList []error
+	nsList := createNamespacesList(phase.NamespaceRange)
+	for _, nsName := range nsList {
+		for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
+			id, err := getIdentifier(&phase.ObjectBundle[j])
+			if err != nil {
+				errList = append(errList, err)
+				continue
+			}
+			instances, exists := ctx.GetState().Get(nsName, id)
+			if !exists {
+				instances = &state.InstancesState{
+					ReplicaCount: 0,
+					Object:       phase.ObjectBundle[j],
+				}
+			}
+			currentInstancesCount := instances.ReplicaCount
+			instances.ReplicaCount = phase.ReplicasPerNamespace
```

Comment:
IIUC this logic is not correct because you're using instances.ReplicaCount as the indicator for 'existing no. of replicas of the object' (let's say 10) and then setting it to new value directly in the next line to the desired future state (let's say 100). So there can be problems like:

- If someone else accesses the state in the meanwhile, it's possible that they end up thinking the current no. of replicas is 100 instead of 10
- If going ahead to create the additional 90 replicas, you fail to create some - then you would end up in an inconsistent state (i.e reality != recorded state)

So IMO, you should increment it by 1 (using a lock) whenever you successfully create a replica. Sth like this:

```
err := CreateReplica()
if err == nil {
  lock.Lock()
  instances.ReplicaCount++
  lock.Unlock()
}
```

Reply:
Btw - Even with the above I think it's still possible to diverge recorded state from real state over time. But I think it's much safer than the above. We may want to fully solve the problem by listing objects at the start of this function to reconcile with correct state - but I'm fine neglecting that for now with a TODO.

Reply:
> If someone else accesses the state in the meanwhile, it's possible that they end up thinking the current no. of replicas is 100 instead of 10

There is an assumption that only one operation can be perform on certain object class (\<namespcae x basename x kind x group\>), so this kind of situation shouldn't ever happen.

> If going ahead to create the additional 90 replicas, you fail to create some - then you would end up in an inconsistent state (i.e reality != recorded state)

In this implementation, state = desired state not real cluster state. At some point we will use discovering cluster state, but not now.